### PR TITLE
[optim, config] feat: add support for Muon optimizer via dion

### PR DESCRIPTION
**Pull request by clarkipeng**
This PR introduces support for the Muon optimizer (Momentum Orthogonalized Optimizer) using the dion backend.
It implements a hybrid optimization strategy where:

Muon is applied to high-dimensional parameters (2D+ matrices), excluding embeddings and the LM head, and with rms-normalization.

AdamW is used as the fallback for 1D tensors (vectors), embeddings, and the LM head.

This implementation requires DeviceMesh for distributed training and adds dion as a project dependency.

### Test
Screenshot 2025-11-24 at 6 59 38 PM

Validated by running qwen3_vl_8b with optimizer="muon" on a 2xH100 node with seq-len = 32784. 

(adamw ooms with same hyperparams)

### API and Usage Example

To use the Muon optimizer, simply update the optimizer argument in your configuration or CLI. No extra configuration is needed for the parameter splitting logic; it is handled automatically.

### Design & Code Changes

This PR modifies how parameter groups are constructed in `build_optimizer` without fsdp2. Before, if the parameter groups were created outside of the ```build_optimizer``` call without fsdp2, weight decay wouldn't be applied. Now, muon (if_enabled) and weight decay are automatically applied to any parameter group passed into ```build_optimizer```.

### CODE REVIEW

**Comment by Luosuu** [comment]
I think in torch 2.9 muon is supported?
https://docs.pytorch.org/docs/stable/generated/torch.optim.Muon.html

**Comment by clarkipeng** [comment]
Oh good point, I'll give it a go

**Comment by clarkipeng** [comment]
@Luosuu Would we want to use torch 2.9? Then we probably need to rewrite many of the optimizations that [dion](https://github.com/microsoft/dion/blob/main/dion/muon.py) does.

Did a quick experiment. If I run a naive MultiOptimizer with muon and adamw the runtime is far worse... (Qwen3-1.7B with 8xH100)
W B Chart 12_13_2025, 1_52_37 AM
W B Chart 12_13_2025, 1_53_02 AM
W B Chart 12_13_2025, 1_53_16 AM

**Comment by Luosuu** [comment]
Interesting, so you mean torch 2.9 muon is numerically same with dion muon but slower?

**Comment by clarkipeng** [comment]
> Interesting, so you mean torch 2.9 muon is numerically same with dion muon but slower?

Yes, particularly in the parallel setting. Dion also implements triton kernels for newton-shutlz so it could be faster than shown above

**Comment by Luosuu** [comment]
I am curious whether the speed diff would still be large for larger models. I thought optimizer step time would not be a very large portion in the end to end training.

VeOmni provides trace profiling by pytorch profiler which can be enabled by `enable_profiling: true`.
Could you please show the trace with dion muon and torch 2.9 muon on a larger model so that we can see the diff more clearly? Thank you very much.

**Comment by clarkipeng** [comment]
Sure!
For reference my 2.9 Muon implementation is the following naive approach
```
optimizers = {}
foreach = not fused

muon_groups = [g for g in param_groups if g.get("algorithm") == "muon"]
optimizers["muon"] = Muon(muon_groups, lr=lr, weight_decay=weight_decay, eps=eps)
            
adamw_groups = [g for g in param_groups if g.get("algorithm") == "adamw"]
optimizers["adamw"] = AdamW(adamw_groups, lr, betas, eps, weight_decay, fused=fused, foreach=foreach)
            
optim = MultiOptimizer(model, optimizers, list(optimizers.keys()))
```

Here are the results on Qwen3-8B, the difference does seem to decrease.

W B Chart 12_13_2025, 11_39_41 PM

# Traces
## dion_muon:
```
  {
    "ph": "X", "cat": "user_annotation", "name": "Optimizer.step#Muon.step", "pid": 1626701, "tid": 1626701,
    "ts": 6394723773641.521, "dur": 125112.171,
    "args": {
      "External id": 145593,"Record function id": 0, "Ev Idx": 47800
    }
  },
  {
    "ph": "X", "cat": "python_function", "name": "dion/muon.py(130): step", "pid": 1626701, "tid": 1626701,
    "ts": 6394723773678.061, "dur": 125013.852,
    "args": {
      "Python parent id": 1591820, "Python id": 1591831, "Ev Idx": 1792869
    }
  },
```
## 2.9_muon:
```
  {
    "ph": "X", "cat": "user_annotation", "name": "Optimizer.step#Muon.step", "pid": 1637316, "tid": 1637316,
    "ts": 6396247796158.178, "dur": 2548154.907,
    "args": {
      "External id": 145593,"Record function id": 0, "Ev Idx": 47800
    }
  },
  {
    "ph": "X", "cat": "python_function", "name": "torch/optim/_muon.py(166): step", "pid": 1637316, "tid": 1637316,
    "ts": 6396247796195.376, "dur": 2547985.927,
    "args": {
      "Python parent id": 1591688, "Python id": 1591699, "Ev Idx": 1935281
    }
  },
```

**Comment by clarkipeng** [comment]
@Luosuu Any thoughts?

**Comment by Luosuu** [comment]
Dion looks cool but I am suspicious about whether Microsoft would maintain it for long term, especially considering that its functionality is overlapped by official pytorch. 
Thank you very much for your report but my personal opinion is to support muon by upgrading to torch 2.9

**Comment by Luosuu** [comment]
I am requesting another reviewer for this for more thoughts on this
